Add DataFetcher and ShardDetector interface changes

Renjuju opened the pull request:
## Changes
* Updated ShardDetector interface exposing method for overriding implementation
* Added DataFetcher interface associated with KinesisDataFetcher to allow customer to pass in custom DataFetcher

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Renjuju committed:
Add DataFetcher and ShardDetector interface changes

Renjuju committed:
Allow passing in custom shard detector and custom data fetcher through configs

Renjuju committed:
update javadoc comment

Renjuju committed:
Add extra type check for KinesisShardDetector

Renjuju committed:
Update javadoc

Renjuju committed:
udpate prefetchRecordsPublisher

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/KinesisShardDetector.java:
```diff
@@ -232,6 +232,11 @@ void cachedShardMap(final List<Shard> shards) {
+    public ListShardsResponse listNextShards(final String nextToken) {
+        return listShards(nextToken);
+    }
+
```
As discussed, this might not be needed.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java:
```diff
@@ -260,7 +263,15 @@ protected Scheduler(@NonNull final CheckpointConfig checkpointConfig,
+                    shardDetector instanceof KinesisShardDetector) {
+                return this.leaseManagementConfig.customShardDetectorProvider().apply(
+                        Pair.of(streamConfig.streamIdentifier(), (KinesisShardDetector) shardDetector));
+            }
```
1. We need not expose our shard detector for customer to provide their custom impl
2. this logic should be moved to shardSyncTaskManagerProvider, which would leverage the customShardDetector in leaseManagementCOnfig and create appropriate shard detector from the DDBLeaseManagementFactory

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementConfig.java:
```diff
@@ -145,6 +148,18 @@
+     *    // pair contains the StreamIdentifier and KinesisShardDetector used by KCL
+     * </pre>
+     */
+    private Function<Pair<StreamIdentifier, KinesisShardDetector>, ShardDetector> customShardDetectorProvider;
```
Let's avoid passing KinesisShardDetector and change this to,
```private Function<StreamConfig, ShardDetector> customShardDetectorProvider;```

StreamConfig should be the only info we should be giving to the client application. If they choose to use KinesisShardDetector as a delegate or a  base class, they should be able to pass the required constructor args themselves.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardDetector.java:
```diff
@@ -32,4 +33,5 @@ default StreamIdentifier streamIdentifier() {
         throw new UnsupportedOperationException("StreamName not available");
     }
 
+    ListShardsResponse listNextShards(final String nextToken);
```
As discussed, let's move this to custom shard detector

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/GetRecordsRetrievalStrategy.java:
```diff
@@ -47,9 +48,9 @@
+     * @return DataFetcher
      */
-    KinesisDataFetcher getDataFetcher();
+    DataFetcher getDataFetcher();
```
We are breaking backward compatibility here
```
public class Test{

    public static void main(String[] args){
        ITest test = new ITest() {
            @Override
            public DataFetcherImpl dataFetcher() {
                return new DataFetcherImpl();
            }
        };

        // Backward compatibility break
        DataFetcherImpl impl = test.dataFetcher();
        System.out.println(impl.add());
    }

//    interface ITest {
//        DataFetcherImpl dataFetcher();
//    }

    // comment
    interface ITest {
        DataFetcher dataFetcher();
    }

//    public static class DataFetcherImpl{
//        public int add() { return 3;}
//    }

    //comment
    public static class DataFetcherImpl implements DataFetcher{
        public int add() { return 3;}
    }

    interface DataFetcher {
        int add();
    }
}
```

ashwing replied:
What we can do here is add another interface method which is 

```
KinesisDataFetcher getDataFetcher();

default Optional<DataFetcher> getDataFetcherOverride() {
 return Option.empty();
}

default DataFetcher dataFetcher() {
 return getDataFetcherOverride().orElse(getDataFetcher());
}
```

Now KCL callers like  `impl.getDataFetcher()` needs to be changed to `impl.dataFetcher()`

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalConfig.java:
```diff
@@ -52,6 +59,18 @@
+     *    // pair contains the StreamIdentifier and KinesisDataFetcher used by KCL
+     * </pre>
+     */
+    private Function<Pair<KinesisDataFetcher, StreamIdentifier>, DataFetcher> customDataFetcherProvider;
```
what we need here is 
```Function<SomeClass(StreamIdentifier, ShardId), DataFetcher> customDataFetcherProvider;```

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalSpecificConfig.java:
```diff
@@ -15,11 +15,27 @@
+     * @return a retrieval factory that can create an appropriate retriever with a given data fetcher
+     */
+    default RetrievalFactory retrievalFactory(Function<Pair<KinesisDataFetcher, StreamIdentifier>, DataFetcher> dataFetcherProvider) {
+        return null;
```
Let's throw UnsupportedOperationException here

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/DataFetcher.java:
```diff
@@ -0,0 +1,44 @@
+import software.amazon.kinesis.retrieval.DataFetcherResult;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+public interface DataFetcher {
```
Let's ensure all the public methods of KinesisDateFetcher is made as interface methods here.

ashwing replied:
Can you resolve this conversation after validating

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/KinesisDataFetcher.java:
```diff
@@ -263,24 +266,43 @@ public void restartIterator() {
+    }
+
+    @Override
+    public GetRecordsRequest getRequest(String nextIterator)  {
```
getGetRecordsRequest

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/KinesisDataFetcher.java:
```diff
@@ -263,24 +266,43 @@ public void restartIterator() {
-        final AWSExceptionManager exceptionManager = createExceptionManager();
-        GetRecordsRequest request = KinesisRequestsBuilder.getRecordsRequestBuilder().shardIterator(nextIterator)
+    @Override
+    public GetRecordsResponse getResponse(GetRecordsRequest request) throws ExecutionException, InterruptedException, TimeoutException {
```
getGetRecordsResponse. Lets throw Exception here, so that the custom impl can choose to throw whatever exception it wants

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/DataFetcher.java:
```diff
@@ -0,0 +1,44 @@
+
+    boolean isShardEndReached();
+
+    GetRecordsResponse getResponse(GetRecordsRequest request) throws ExecutionException, InterruptedException, TimeoutException;
+
+    GetRecordsRequest getRequest(@NonNull String nextIterator) throws ExecutionException, InterruptedException, TimeoutException;
```
Does these types GetRecordsRequest and GetRecordsResponse fit for archiver read only client as well? If not then we need generics here,

```
public interface DataFetcher<Req, Resp> {
...
Resp getResponse(Req request);
Req getRequest(String nextIterator);
Resp getRecords(String nextIterator);
...
}
```

check if we need to do this for ```String getNextIterator(GetShardIteratorRequest request)``` as well

Renjuju replied:
I created methods to convert the requests and responses to types that Archiver can use

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PollingConfig.java:
```diff
@@ -102,9 +105,18 @@
+    @Override
+    public RetrievalFactory retrievalFactory(
+            Function<Pair<KinesisDataFetcher, StreamIdentifier>, DataFetcher> dataFetcherProvider
+    ) {
```
indent fix

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PollingConfig.java:
```diff
@@ -102,9 +105,18 @@
     public RetrievalFactory retrievalFactory() {
         return new SynchronousBlockingRetrievalFactory(streamName(), kinesisClient(), recordsFetcherFactory,
-                maxRecords(), kinesisRequestTimeout);
+                maxRecords(), kinesisRequestTimeout, null);
```
Let's have a dedicated constructor and not pass null here

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -47,17 +49,25 @@
     @Deprecated
     public SynchronousBlockingRetrievalFactory(String streamName, KinesisAsyncClient kinesisClient, RecordsFetcherFactory recordsFetcherFactory, int maxRecords) {
-        this(streamName, kinesisClient, recordsFetcherFactory, maxRecords, PollingConfig.DEFAULT_REQUEST_TIMEOUT);
+        this(streamName, kinesisClient, recordsFetcherFactory, maxRecords, PollingConfig.DEFAULT_REQUEST_TIMEOUT, null);
```
refer prev comment to avoid passing null

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -66,8 +76,17 @@ public GetRecordsRetrievalStrategy createGetRecordsRetrievalStrategy(@NonNull fi
                 StreamIdentifier.singleStreamInstance(streamName);
-        return new SynchronousGetRecordsRetrievalStrategy(
-                new KinesisDataFetcher(kinesisClient, streamIdentifier, shardInfo.shardId(), maxRecords, metricsFactory, kinesisRequestTimeout));
+        KinesisDataFetcher kinesisDataFetcher = new KinesisDataFetcher(kinesisClient,
```
This might be a wasted object. Move to a method and call in the next ternary statement

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -66,8 +76,17 @@ public GetRecordsRetrievalStrategy createGetRecordsRetrievalStrategy(@NonNull fi
+                metricsFactory,
+                kinesisRequestTimeout);
+
+        DataFetcher dataFetcher = this.dataFetcherProvider == null ? kinesisDataFetcher :
```
final. Use final wherever applicable

ashwing reviewed (commented):
Initial comments

Commit:
ShardDetector now passed created in DynamoDBLeaseManagementFactory
DataFetcher interface updated
Updated Functional interface for ShardDetector and DataFetcher

Commit:
update var names

Commit:
Added DataFetcherProviderConfig to make functional interface arugments clear

Commit:
Deprecate getDataFetcher method

Commit:
Add additional vars for DataFetcherProviderConfig

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -365,7 +378,8 @@ public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient, fi
      * @param dynamoDbRequestTimeout
      * @param billingMode
      */
```
Let's update the doc

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -415,7 +431,8 @@ public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient, fi
      * @param billingMode
      * @param leaseSerializer
```
Let's update doc

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -517,13 +536,19 @@ public ShardDetector createShardDetector() {
-                cacheMissWarningModulus, dynamoDbRequestTimeout);
+        if (this.customShardDetectorProvider != null) {
+            return this.customShardDetectorProvider.apply(streamConfig);
+        } else {
```
nit : we don't need else

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/DataFetcherProviderConfig.java:
```diff
@@ -0,0 +1,26 @@
+import software.amazon.kinesis.metrics.MetricsFactory;
+
+
+/**
```
Need license blob for all the new files

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/DataFetcherProviderConfig.java:
```diff
@@ -0,0 +1,26 @@
+ * Configuration needed for custom data fetchers
+ */
+@Data
+public class DataFetcherProviderConfig {
```
Since this is a class exposed to end customer, we should have the flexibility of adding new fields. Let's make this an interface and expose to customers.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/GetRecordsRetrievalStrategy.java:
```diff
@@ -41,15 +43,24 @@
+    @Deprecated
     KinesisDataFetcher getDataFetcher();
+
+    default Optional<DataFetcher> getDataFetcherOverride() {
```
Let's add comment

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/GetRecordsRetrievalStrategy.java:
```diff
@@ -41,15 +43,24 @@
+        return Optional.empty();
+    }
+
+    default DataFetcher dataFetcher() {
```
Let's add doc comment

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalConfig.java:
```diff
@@ -52,6 +57,11 @@
+    /**
+     * Configurable functional interface to override the existing DataFetcher.
+     */
+    private Function<DataFetcherProviderConfig, DataFetcher> customDataFetcherProvider;
```
DataFetcherProviderConfig should be an interface as per previous comment

ashwing replied:
Going through this again, why this shouldn't be passed directly as PollingConfig as we don't need datafetcher for fanoutconfig. This way we can avoid the new interface method in RetrievalSpecificConfig and changes in retrievalFactory() of this class. Also, we can prevent customers from setting datafetcher in fanout mode.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/DataFetcher.java:
```diff
@@ -0,0 +1,61 @@
+import software.amazon.kinesis.retrieval.DataFetcherResult;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+public interface DataFetcher {
```
Need license blob for new files

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/DataFetcher.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    void restartIterator();
+
+    void resetIterator(String shardIterator, String sequenceNumber, InitialPositionInStreamExtended initialPositionInStream);
```
let's add doc comments for all the methods

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/DataFetcher.java:
```diff
@@ -0,0 +1,61 @@
+
+    GetRecordsResponse getRecords(@NonNull String nextIterator);
+
+    software.amazon.kinesis.common.StreamIdentifier getStreamIdentifier();
```
Let's move this to import statements

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -47,14 +49,34 @@
+                                               int maxRecords,
+                                               Duration kinesisRequestTimeout,
+                                               Function<DataFetcherProviderConfig, DataFetcher> dataFetcherProvider) {
+        this.streamName = streamName;
+        this.kinesisClient = kinesisClient;
+        this.recordsFetcherFactory = recordsFetcherFactory;
+        this.maxRecords = maxRecords;
+        this.kinesisRequestTimeout = kinesisRequestTimeout;
```
replace these lines with with this(s,k,r,m,k)

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -66,8 +88,19 @@ public GetRecordsRetrievalStrategy createGetRecordsRetrievalStrategy(@NonNull fi
+                maxRecords,
+                metricsFactory,
+                kinesisRequestTimeout) :
+                this.dataFetcherProvider.apply(new DataFetcherProviderConfig(streamIdentifier,
+                        shardInfo.shardId(),
+                        metricsFactory,
+                        maxRecords));
```
DataFetcherProviderConfig can now be an arg for KinesisDataFetcher as well

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -66,8 +88,19 @@ public GetRecordsRetrievalStrategy createGetRecordsRetrievalStrategy(@NonNull fi
+                shardInfo.shardId(),
+                maxRecords,
+                metricsFactory,
+                kinesisRequestTimeout) :
```
why kinesisRequestTimeout missing from DataFetcherProviderConfig

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousGetRecordsRetrievalStrategy.java:
```diff
@@ -45,9 +46,20 @@ public void shutdown() {
+    @Override
+    public KinesisDataFetcher getDataFetcher() {
+        // returns null as dataFetcher() will now return the current dataFetcher
+        return null;
```
Let's throw UnsupportedOperationException if we expect this not to be called

ashwing reviewed (commented):
shaping up well. Added comments

Renjuju committed:
Add missing doc strings & use interface for Config impl

Renjuju commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/DataFetcherProviderConfig.java:
```diff
@@ -1,26 +1,48 @@
 import software.amazon.kinesis.common.StreamIdentifier;
 import software.amazon.kinesis.metrics.MetricsFactory;
 
+public interface DataFetcherProviderConfig {
```
@ashwing 
Should we expose the DataFetcherProviderConfig and have that be used instead of the current KinesisDataFetcherProviderConfig if DataFetcherProviderConfig is provided by the user?

ashwing replied:
we will expose only the interface DataFetcherProviderConfig to the end user. The actual concrete class we give to the customer doesn't matter to them.

Renjuju committed:
Fix test

Renjuju committed:
tests & readability

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -109,22 +111,24 @@
                 maxListShardsRetryAttempts, maxCacheMissesBeforeReload, listShardsCacheAllowedAgeInSeconds,
                 cacheMissWarningModulus, TableConstants.DEFAULT_INITIAL_LEASE_TABLE_READ_CAPACITY,
-                TableConstants.DEFAULT_INITIAL_LEASE_TABLE_WRITE_CAPACITY);
+                TableConstants.DEFAULT_INITIAL_LEASE_TABLE_WRITE_CAPACITY, customShardDetectorProvider);
```
As a rule of thumb, we should not modify or touch the existing and deprecated public constructor signatures. We should avoid passing Function<StreamConfig, ShardDetector> customShardDetectorProvider in this constructor args. The only change we should see is a new constructor with customShardDetectorProvider, and rest of the constructors should pass null as their last argument.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/events/ProcessRecordsInput.java:
```diff
@@ -23,6 +23,7 @@
 import lombok.Getter;
 import lombok.ToString;
 import lombok.experimental.Accessors;
+import software.amazon.awssdk.services.kinesis.model.Record;
```
Additional import

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalConfig.java:
```diff
@@ -117,17 +127,20 @@ public RetrievalConfig initialPositionInStreamExtended(InitialPositionInStreamEx
             retrievalFactory = retrievalSpecificConfig.retrievalFactory();
         }
+
+        if (retrievalSpecificConfig != null && customDataFetcherProvider != null) {
```
At this point the retrieval specific config will definitely not be null. Since customDataFetcherProvider is only applicable for PollingConfig, what you need here is

```
    public RetrievalFactory retrievalFactory() {

        if (retrievalFactory == null) {
            if (retrievalSpecificConfig == null) {
                retrievalSpecificConfig = new FanOutConfig(kinesisClient())
                        .applicationName(applicationName());
                retrievalSpecificConfig = appStreamTracker.map(multiStreamTracker -> retrievalSpecificConfig,
                        streamConfig -> ((FanOutConfig)retrievalSpecificConfig).streamName(streamConfig.streamIdentifier().streamName()));
            }
            if(customDataFetcherProvider != null) {
                retrievalFactory = retrievalSpecificConfig.retrievalFactory(customDataFetcherProvider);
            } else {
                retrievalFactory = retrievalSpecificConfig.retrievalFactory();
            }
        }
        return retrievalFactory;
    }
```

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalSpecificConfig.java:
```diff
@@ -15,11 +15,24 @@
+     * @return a retrieval factory that can create an appropriate retriever with a given data fetcher
+     */
+    default RetrievalFactory retrievalFactory(Function<DataFetcherProviderConfig, DataFetcher> dataFetcherProvider) {
+        throw new UnsupportedOperationException();
```
Can you add an exception message here?

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/KinesisDataFetcher.java:
```diff
@@ -97,11 +96,29 @@ public KinesisDataFetcher(KinesisAsyncClient kinesisClient, StreamIdentifier str
         this.streamAndShardId = streamIdentifier.serialize() + ":" + shardId;
```
Since this constructor is introduced as part of multistreaming story, we can remove this and we can use your new constructor

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -47,14 +49,30 @@
-    public SynchronousBlockingRetrievalFactory(String streamName, KinesisAsyncClient kinesisClient, RecordsFetcherFactory recordsFetcherFactory, int maxRecords, Duration kinesisRequestTimeout) {
+    private Function<DataFetcherProviderConfig, DataFetcher> dataFetcherProvider;
+
+    public SynchronousBlockingRetrievalFactory(String streamName,
```
Mark as deprecated?

Renjuju replied:
Would this be marked as deprecated? `PollingConfig.java` still uses the existing constructor to create a retrievalFactory when a dataFetcherProvider is not passed in

ashwing replied:
We need to maintain one non-deprecated constructor for public , so that we can extend it without branching in a longer run. We can pass the kinesis specific data fetcher provider from PollingConfig.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousBlockingRetrievalFactory.java:
```diff
@@ -66,8 +84,18 @@ public GetRecordsRetrievalStrategy createGetRecordsRetrievalStrategy(@NonNull fi
-        return new SynchronousGetRecordsRetrievalStrategy(
-                new KinesisDataFetcher(kinesisClient, streamIdentifier, shardInfo.shardId(), maxRecords, metricsFactory, kinesisRequestTimeout));
+
+        final DataFetcherProviderConfig kinesisDataFetcherProviderConfig = new KinesisDataFetcherProviderConfig(streamIdentifier,
+                shardInfo.shardId(),
+                metricsFactory,
+                maxRecords,
+                kinesisRequestTimeout);
```
nit: possible wasted object. move this to a private method getKinesisDataFetcherProviderConfig() and call this.dataFetcherProvider.apply(getKinesisDataFetcherProviderConfig());

Renjuju replied:
Not sure if it will ever be wasted in this context. The next line uses kinesisDataFetcherProviderConfig in both parts of the ternary statement

ashwing replied:
Ya you are right. Ignore this then.

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousGetRecordsRetrievalStrategy.java:
```diff
@@ -45,9 +46,20 @@ public void shutdown() {
+    public DataFetcher dataFetcher() {
         return dataFetcher;
     }
+
+    @Override
+    public KinesisDataFetcher getDataFetcher() {
+        // returns null as dataFetcher() will now return the current dataFetcher
+        throw new UnsupportedOperationException();
+    }
```
You dont need this if you are directly overriding dataFetcher()

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/SynchronousGetRecordsRetrievalStrategy.java:
```diff
@@ -45,9 +46,20 @@ public void shutdown() {
+
     @Override
-    public KinesisDataFetcher getDataFetcher() {
+    public Optional<DataFetcher> getDataFetcherOverride() {
+        return Optional.of(dataFetcher);
+    }
```
You dont need this if you are directly overriding dataFetcher()

ashwing reviewed (commented):
Left comments for refactoring

Renjuju committed:
Remove constructor args from DDBLeaseManagementFactory
Code clean up
Update RetrievalConfig

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -454,6 +457,78 @@ public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient,
+     * @param leaseSerializer
+     * @param customShardDetectorProvider
+     */
+    public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient,
```
Let's update this as discussed offline

ashwing reviewed (commented):
Comments on public facing configuration and classes.

Renjuju committed:
deprecate method, remove extra constructor

Renjuju committed:
Move dataFetcher to PollingConfig, add validation

Renjuju committed:
remove override method

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalConfig.java:
```diff
@@ -117,17 +122,29 @@ public RetrievalConfig initialPositionInStreamExtended(InitialPositionInStreamEx
+                        streamConfig.streamIdentifier() == null || streamConfig.streamIdentifier().streamName() == null);
+
+        if(isInvalidPollingConfig) {
+            throw new UnsupportedOperationException("Invalid config: multistream enabled with streamName or single stream with no streamName");
```
IllegalArgumentException

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PollingConfig.java:
```diff
@@ -35,13 +36,23 @@
+    /**
+     * @param kinesisClient Client used to access Kinesis services.
+     */
+    public PollingConfig(KinesisAsyncClient kinesisClient) {
```
Given that we have explicit constructors, we need to remove lombok's @RequiredArgsConstructor from @Data (@Data - @RequiredArgsConstructor) = @Getter, @Setter, @ToString, @EqualsAndHashCode

Renjuju replied:
Looks like adding in explicit constructors removes the RequiredArgsConstructor - tested this + https://github.com/rzwitserloot/lombok/issues/1269

Renjuju committed:
remove unused annotation + update exception type

Renjuju committed:
make lombok annotations explicit

Renjuju committed:
remove commented code

Renjuju committed:
Clean up SynchronousBlockingRetrievalFactory

ashwing commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java:
```diff
@@ -452,6 +457,7 @@ public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient,
         this.dynamoDbRequestTimeout = dynamoDbRequestTimeout;
         this.billingMode = billingMode;
         this.leaseSerializer = leaseSerializer;
+        this.customShardDetectorProvider = customShardDetectorProvider;
```
I don't see an usage of this field in this class

ashwing reviewed (commented):
Can you check the usage of sharddetector provider in DynamoDBLeaseManagementFactory

Renjuju committed:
use custom shard detector

Renjuju committed:
update doc strings

ashwing reviewed (approved):
LGTM
